addons/sale/models/product_template.py

```diff
@@ -118,3 +118,24 @@ def get_import_templates(self):
                     'template': '/product/static/xls/product_template.xls'
                 }]
         return res
+
+    @api.model
+    def _get_incompatible_types(self):
+        return []
+
+    @api.constrains(lambda self: self._get_incompatible_types())
+    def _check_incompatible_types(self):
+        incompatible_types = self._get_incompatible_types()
+        if len(incompatible_types) < 2:
+            return
+        fields = self.env['ir.model.fields'].sudo().search_read(
+            [('model', '=', 'product.template'), ('name', 'in', incompatible_types)],
+            ['field_description'])
+        field_descriptions = [v['field_description'] for v in fields]
+        field_list = incompatible_types + ['name']
+        values = self.read(field_list)
+        for val in values:
+            incompatible_values = [val[f] for f in incompatible_types if val[f]]
+            if len(incompatible_values) > 1:
+                raise ValidationError(_("The product (%s) has incompatible values: %s",
+                                        val['name'], ','.join(field_descriptions)))
```

Comment:
```suggestion
        field_descriptions = {v['name']: v['field_description'] for v in fields}
        field_list = incompatible_types + ['name']
        values = self.read(field_list)
        for val in values:
            incompatible_fields = [f for f in incompatible_types if val[f]]
            if len(incompatible_fields) > 1:
                raise ValidationError(_(
                     "The product (%s) has incompatible values: %s",
                     val['name'], 
                     ,'.join(field_descriptions[v] for v in incompatible_fields),
                 ))
```

I know that we're only targeting rental & subscription products right now, so we won't see the difference, but if a third incompatible type was added, you would still print all the incompatible types in the error message (even if it was not set on the current product).